sidecar/resources.py

```diff
@@ -0,0 +1,141 @@
+    print_prefix = "Thread: " if thread else ""
+    v1 = client.CoreV1Api()
+    namespace = os.getenv("NAMESPACE", current)
+    if namespace == "ALL":
+        stream = watch.Watch().stream(getattr(v1, _list_for_all_namespaces[resource]))
+    else:
+        stream = watch.Watch().stream(getattr(v1, _list_namespaced[resource]), namespace=namespace)
+
+    for event in stream:
+        destFolder = targetFolder
+        metadata = event['object'].metadata
+        if metadata.labels is None:
+            continue
+        print(f'{print_prefix}Working on {resource} {metadata.namespace}/{metadata.name}')
+        if label in event['object'].metadata.labels.keys():
+            print(f"{print_prefix}{resource} with label found")
+            if event['object'].metadata.annotations is not None:
+                if folderAnnotation in event['object'].metadata.annotations.keys():
+                    destFolder = event['object'].metadata.annotations[folderAnnotation]
+                    print(f'{print_prefix}ound a folder override annotation, '
+                          'placing the {resource} in: {destFolder}')
+            dataMap = event['object'].data
+            if dataMap is None:
+                print(f"{print_prefix}{resource} does not have data.")
+                continue
+            eventType = event['type']
+            for data_key in dataMap.keys():
+                print(f"{print_prefix}File in {resource} {data_key} {eventType}")
+
+                if (eventType == "ADDED") or (eventType == "MODIFIED"):
+                    filename, filedata = _get_file_data_and_name(data_key, dataMap[data_key],
+                                                                 resource)
+                    writeTextToFile(destFolder, filename, filedata)
+
+                    if url is not None:
+                        request(url, method, payload)
+                else:
+                    filename = data_key[:-4] if data_key.endswith(".url") else data_key
+                    removeFile(destFolder, filename)
+                    if url is not None:
+                        request(url, method, payload)
+
+
+def _watch_resource_loop(*args):
+    try:
+        _watch_resource_iterator(*args)
+    except ApiException as e:
+        if e.status != 500:
+            print(f"ApiException when calling kubernetes: {e}\n")
+        else:
+            raise
+    except ProtocolError as e:
+        print(f"ProtocolError when calling kubernetes: {e}\n")
+    except Exception as e:
+        print(f"Received unknown exception: {e}\n")
+
+
+def watchForChanges(label, targetFolder, url, method, payload,
+                    current, folderAnnotation, resources):
+
+    if len(resources) == 2:
+        Thread(target=_watch_resource_loop,
+               args=(label, targetFolder, url, method, payload,
+                     current, folderAnnotation, resources[1], True)
+               ).start()
```

Comment:
Will an exception in the thread be raised to the main thread too? If not we would not have a change of restarting if that thread fails or?

Reply:
Good point. That is why I put the error check inside the loop.

But if we get a 500 error it is raised and dies... I will think about it this evening a bit and put in a change (I am thinking about checking in the main thread if the loop thread is still running and raising there otherwise...)